### PR TITLE
Dev 1.0

### PR DESCRIPTION
**Pull request by ajoshdee**
User Profile:
- Login/Logout
- Edit Profile
- View own thread

Thread/Comment:
- Create/Remove/Edit (owner)
- Upvote/downvote thread
- Can view dates when thread/comment is added/updated.
- Pagination
- Search thread(low priority)
- Sort trending threads(most likes & new)

### CODE REVIEW

File: app/facebook.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            data={'code': request.args['code'],
+                  'grant_type': 'authorization_code',
+                  'redirect_uri': self.get_callback_url()}
+        )
+        me = oauth_session.get('me?fields=id,email,name').json()
+        return (
+            'facebook$' + me['id'],
+            me.get('name'),  # Facebook does not provide
+                                            # username, so the email's user
+                                            # is used instead
```
Comment (maragatchalian) [inline app/facebook.py]: Please change the format of comments in L64-L66.
You may refer to [here](https://www.python.org/dev/peps/pep-0008/#comments)

File: app/facebook.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.provider_name = provider_name
+        credentials = current_app.config['OAUTH_CREDENTIALS'][provider_name]
+        self.consumer_id = credentials['id']
+        self.consumer_secret = credentials['secret']
+
+    def authorize(self):
+        pass
+
+    def callback(self):
+        pass
```
Comment (maragatchalian) [inline app/facebook.py:18]: L14 - L18:
Are these functions used po? :bow:

**Comment by josefmonje** [comment]
- request method checking
- creating/saving models
- change state on GET request
- separation of app modules

---

```
def register(request):
    form = RegistrationForm(request.POST)
    if request.method == 'POST' and form.validate():
        user = User()
        user.username = form.username.data
        user.email = form.email.data
        user.save()
        redirect('register')
    return render_response('register.html', form=form)
```

> _example code from WTForms [docs](https://wtforms.readthedocs.org/en/latest/crash_course.html#using-forms)_

```
from flask import request

...
    if request.method == 'POST' and form.validate():
...
```

check request method if multiple methods are available in view

---

```
        post = Post(title=form.post.data, date_created=datetime.utcnow(), author=g.user)
        db.session.add(post)
        db.session.commit()
```

> _submitted code for creating and saving models_

above code is repeated within each view
suggestion is for the models to have a save() method

```
class Post(db.Model):

    ...

    def __init__(self, *args, **kwargs):
        ...


    def save(self, *args, **kwargs):
        ...
        db.session.commit()
```

> _model_

this will allow you to instantiate models with parameters. this class also provides a .save() method for saving model instances.

```
post = Post(...)
post.save()
```

> _view_

sample usage

---

```
@app.route('/delete/post/<int:id>')
@login_required
def delete_post(id):
    post = Post.query.get(id)
    if post is None:
        flash('Post not found.')
        return redirect(url_for('index'))
    if post.author.id != g.user.id:
        flash('You cannot delete this post.')
        return redirect(url_for('index'))
    db.session.delete(post)
    db.session.commit()
    flash('Your post has been deleted.')
    return redirect(url_for('index'))
```

> _sample delete post url on a GET request_

avoid changing state on GET request. let's say the person clicked by mistake or a crawler reached this url, it will delete the post. additionally, you might also want to implement soft delete.

---

you might also want to separate your app into various subfolders. ex, user, post, comment then put user related functionality in the respective folder. this will help you keep your code easy to navigate and avoid files with unrelated content.

File: app/facebook.py (hunk 2 above)
Comment (josefmonje) [inline app/facebook.py:18]: it's implemented in `class FacebookSignIn(OAuthSignIn)` :)

File: app/facebook.py (hunk 1 above)
Comment (ajoshdee) [inline app/facebook.py]: Comment removed. Comment no longer applicable.

**Comment by ajoshdee** [comment]
RE: request method checking
Flask-WTF has a function named `validate_on_submit` that checks if the request is a 'POST' request and validates the data submitted. 

image

http://flask-wtf.readthedocs.org/en/latest/quickstart.html

**Comment by ajoshdee** [comment]
-creating/saving models ----- DONE
-change state on GET request ----- DONE

**Pull request merged**